### ISSUE TITLE
Web Assembly (Wasm) Support

### ISSUE THREAD

**Issue by Mohsens22** [realm/realm-dotnet#2199]
Hi.
Does this client library work on Dotnet Web Assembly projects (Such as Uno or Blazor)?
If not, will there be a plan to add the support?

**Comment by nirinchev** [comment]
We don't support Wasm yet, although this is an area of acute interest. The Core Database team is exploring various approaches to get the native bits running in Wasm, but that effort is in its infancy and we're not in a position to make any commitments as to when and if the .NET SDK will be able tor un in Blazor/Uno.

**Comment by Mohsens22** [comment]
Any news on this one?

**Comment by nirinchev** [comment]
Unfortunately, nothing to report here.

**Comment by Mohsens22** [comment]
I believe that flutter has a wasm head so if we don't have wasm compilation on realm core, does it mean that we also can't use it for flutter web?

**Comment by nirinchev** [comment]
That is correct - none of the Realm database SDKs support web. If you're using MongoDB Realm (i.e. Sync), you can use the GraphQL API via http to query data in Atlas, but you can't use the local database.

**Comment by moritz-mg** [comment]
Late to the party, as always: Unity WebGL support is a super important feature for us, really (and a bit of a dealbreaker if not coming in the near future).

Use case: We're designing a Web3 & Mobile game we plan to scale to ~60-70k concurrent users, and are looking for ways to not have to deal with writing and maintaining a bridge from Firebase Web to Unity for WebGL/WASM builds (while we could use the pretty decent mobile SDKs otherwise).

Realm has a vastly superior development paradigm in-engine that made it a very sweet outlook, which makes it extremely attractive, and also has native desktop support, which is a huge plus. In-Browser execution is still a must for our use case at this time.

There's perhaps a way to feed the native Cpp code into whatever compiles Cpp to WASM in Unity. Even though Unity Technologies just started their half-baked own backend-as-a-service, I reckon they'd be super keen on a solution as slick as Realm Sync to support their entire product.

In fact, Realm Sync isn't as important on WebGL/WASM, though my vision for a game you could pick up on your phone as you leave the house and continue playing exactly where you left off in the browser now sadly crumbled (a little bit). :)

**Comment by timahrentlov** [comment]
.NET 7  is not too far of, and WASM is going to be important in that release.
Would be nice to experiment with blazor + realm soon :)

**Comment by connde** [comment]
Hi, any updates on that? We're evaluating the use of Realm for Blazor on multiple enterprise apps, is this on the roadmap?

**Comment by kyurkchyan** [comment]
Just another comment on how much I would love to see realm in Web Assembly. We have multiple projects using realm, and the only option to access realm DB from the web app is through querying the MongoDB database either through Graph QL APIs, or through a custom API layer.

**Comment by nirinchev** [comment]
@kyurkchyan are you looking for database support in WASM or just an API to do one-off queries of MongoDB using C#? We're considering extracting the authentication, functions, and remote mongodb client functionalities from the SDK into a separate package, that will be WASM-compatible, but I'd be curious to hear if that's something that will solve your use case.

**Comment by kyurkchyan** [comment]
Hey @nirinchev. We are looking for DB support in WASM/Blazor. I asked this question about three years ago in [StackOverflow](https://stackoverflow.com/questions/64837344/does-realm-database-support-blazor-web-assembly), and you had a reply to it. I have been closely following this topic since then regularly. 

In most apps we develop, we would like to have as much as possible feature parity and code sharing between mobile and web. Otherwise, we must completely duplicate the core business logic for synced and non-synced scenarios. When working in a sync realm scenario, we heavily rely on the reactive aspects of the realm, and the app is built with reactive extensions. Thus, if the web app can't do that, we will need to reimplement the web app with a more traditional server/client API-based approach, which translates into a different business logic, and in many cases, a different UX. For instance, in WEB API-based approach you need to start thinking about data pagination, which is not an issue in a synced realm scenario.

Thus, we should have synced DB in WASM in the ideal world scenario. Before reaching that ideal world, having shared remote and auth functionality is still a significant gain, as it will simplify the setup. We could go without an explicit API server and talk to MongoDB directly through the auth/API layer built into the realm SDK.

**Comment by nirinchev** [comment]
Thanks, this makes sense! We're still slowly working towards database support for WASM, but there are many non-trivial obstacles, so we still have some way to go before it's feasible to use as a fully supported platform.

Extracting a fully managed implementation of a App Services client is more feasible in the short term and we'll take your feedback into consideration when reviewing the backlog of tasks.

**Comment by sumowesley** [comment]
> @kyurkchyan are you looking for database support in WASM or just an API to do one-off queries of MongoDB using C#? We're considering extracting the authentication, functions, and remote mongodb client functionalities from the SDK into a separate package, that will be WASM-compatible, but I'd be curious to hear if that's something that will solve your use case.

I'd be interested to know if this idea has got much further. We're writing mobile app with Realm but because of app store rules, we have to manage subscriptions outside of the Apps.  We would like to do this through a small Blazor WASM project.

**Comment by nirinchev** [comment]
Yes, that's still something we're interested in and are slowly moving in that direction, but don't have a timeline we can share yet. If you have timeframes for when you'd need this, definitely reach out to your account executive and share those with them as their input has meaningful impact on the engineering prioritization.

**Comment by sumowesley** [comment]
Hi @nirinchev, thank you for the reply. I'm not sure we have an account executive as such. We've been working on a free database for several months looking at how best to structure Realm/MongoDB for our needs.
